# Overview

## **This project is in progress. It's not guaranteed to function properly.**

Altai (**Al**lele-specific **T**ranscript **A**ssembly **I**nstrument) is a reference-based allele-specific transcript assembler. It incorporates variations, e.g. SNPs in a vcf file, in the transcript assembly to assemble the full-length transcript sequences by assigning those variations to their correct phases, i.e. alleles. Its output transcript sequences are expected to be the actual allele-specific sequences rather than subsequences of the reference genome.

Part of Altai is developed based on [Scallop](https://github.com/Kingsford-Group/scallop) ([license](https://github.com/Kingsford-Group/scallop/blob/master/LICENSE)).

# Installation

Altai uses additional libraries of Boost, htslib, and Gurobi. If they have not been installed in your system, you first need to download and install them.

## Download Boost

If Boost has not been downloaded/installed, download Boost [(license)](http://www.boost.org/LICENSE_1_0.txt) from (www.boost.org).
Uncompress it somewhere (compiling and installing are not necessary).

## Install htslib

If htslib has not been installed, download htslib [(license)](https://github.com/samtools/htslib/blob/develop/LICENSE) from (www.htslib.org/) with version 1.5 or higher. Note that htslib relies on zlib. So if zlib has not been installed in your system, you need to install zlib first. 

Use the following commands to build htslib:

```
./configure --disable-bz2 --disable-lzma --disable-gcs --disable-s3 --enable-libcurl=no
make
make install
```

The default installation location of htslib is `/usr/lib`. If you would install it to a different location, replace the above `configure` line with the following (by adding `--prefix=/path/to/your/htslib` to the end):

```
./configure --disable-bz2 --disable-lzma --disable-gcs --disable-s3 --enable-libcurl=no --prefix=/path/to/your/htslib
```

In this case, you also need to export the runtime library path (note that there is an additional `lib` following the installation path):

```
export LD_LIBRARY_PATH=/path/to/your/htslib/lib:$LD_LIBRARY_PATH
```

## Install Gurobi Solver

Gurobi is an optimization solver. If Gurobi has not been installed, download Gurobi Optimizer and corresponding README.txt from (www.gurobi.com/downloads/) with version 9.1.1 or higher. Request a license as you need. [Academic license](https://www.gurobi.com/academia/academic-program-and-licenses/) is free. Follow the instructions to install Gurobi Optimizer and store the license file.

After installation, respectively export Gurobi home directory and license file to environment variable `GUROBI_HOME` and `GRB_LICENSE_FILE`:

```
export GUROBI_HOME=/path/to/Gurobi/dir/linux64
export GRB_LICENSE_FILE=/path/to/Gurobi/license/gurobi.lic
```

## Compile Altai

Use the following to compile Altai:

```
./configure --with-htslib=/path/to/your/htslib --with-boost=/path/to/your/boost
make
```

If some of the dependencies are installed in the default system directory (for example, `/usr/lib`),
then the corresponding `--with-` option might not be necessary. The executable file `altai` will appear at `src/altai`.

# Usage

The usage of `altai` is:

```
./altai -i <input.bam> -j <variants.vcf> [-G <genome.fa>] -o <output-prefix> [options]
```

If you would like to output the transcript sequences, `-G <genome.fa>` is necessary. Otherwise, it's optional.

The `input.bam` is the read alignment file generated by some RNA-seq aligner, (for example, TopHat2, STAR, or HISAT2).
Make sure that it is sorted; otherwise run `samtools` to sort it:

```
samtools sort input.bam > input.sort.bam
```

The reconstructed allele-specific transcripts shall be written as gtf format into `output-prefix.gtf`. Their sequences will be written as fasta format into `output-prefix.fa`.

